src/server/game/Entities/AreaTrigger/AreaTrigger.cpp

```diff
@@ -100,14 +148,256 @@ void AreaTrigger::Update(uint32 p_time)
         Remove(); // expired
 
     WorldObject::Update(p_time);
+    _timeSinceCreated += p_time;
+
+    sScriptMgr->OnAreaTriggerEntityUpdate(this, p_time);
+
+    switch (GetTemplate()->GetType())
+    {
+        case AREATRIGGER_TYPE_SPHERE:
+            SearchUnitInSphere();
+            break;
+        case AREATRIGGER_TYPE_BOX:
+            SearchUnitInBox();
+            break;
+        case AREATRIGGER_TYPE_POLYGON:
+            SearchUnitInPolygon();
+            break;
+        case AREATRIGGER_TYPE_CYLINDER:
+            SearchUnitInCylinder();
+            break;
+    }
 }
 
 void AreaTrigger::Remove()
 {
     if (IsInWorld())
     {
+        if (Unit* caster = GetCaster())
+            caster->_UnregisterAreaTrigger(this);
+
+        for (ObjectGuid insideUnitGuid : _insideUnits)
+            if (Unit* leavingUnit = ObjectAccessor::GetUnit(*this, insideUnitGuid))
+                sScriptMgr->OnAreaTriggerEntityUnitExit(this, leavingUnit);
+
+        sScriptMgr->OnAreaTriggerEntityRemove(this);
+
         SendObjectDeSpawnAnim(GetGUID());
         RemoveFromWorld();
         AddObjectToRemoveList();
     }
 }
+
+void AreaTrigger::SearchUnitInSphere()
+{
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::SearchUnitInBox()
+{
+    float extentsX = GetTemplate()->BoxDatas.Extents[0];
+    float extentsY = GetTemplate()->BoxDatas.Extents[1];
+    float extentsZ = GetTemplate()->BoxDatas.Extents[2];
+
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    float halfExtentsX = extentsX / 2.0;
```

Comment:
The implementation of the box seems legit however G3D offers this implementation already through `AABox`, you could use:

```c++
AABox box({x_low, y_low, z_low}, {x_high, y_high, z_high});
...
box.contains({x, y, z});
```

instead.